### PR TITLE
Regenerate checkpoint states on demand

### LINKED ISSUE #2292: Limit checkpoint states kept in memory
### Description
Relates to #1976 - In addition to limiting the `Store.block_states` collection, we should also limit the number of states kept in `Store.checkpoint_states`.  We should be able to use a strategy very similar to that implemented in #1976 .

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Reduce memory requirements during non-finalised periods by limiting the number of checkpoint states that are held and regenerating them on demand.

## Fixed Issue(s)
fixes #2292

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: storage/src/main/java/tech/pegasys/teku/storage/store/Store.java [hunk 1]
```diff
@@ -390,21 +379,49 @@ public BeaconState getBlockState(Bytes32 blockRoot) {
...
-  public boolean containsCheckpointState(Checkpoint checkpoint) {
-    readLock.lock();
+  private BeaconState regenerateCheckpointState(
+      final Checkpoint checkpoint, Function<Bytes32, BeaconState> getBlockState) {
     try {
-      return checkpoint_states.containsKey(checkpoint);
-    } finally {
-      readLock.unlock();
+      final BeaconState baseState = getBlockState.apply(checkpoint.getRoot());
+      if (baseState == null || baseState.getSlot().equals(checkpoint.getEpochStartSlot())) {
```
Comment (cemozerr) [inline storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:423]: Is this supposed to be:
```suggestion
      if (baseState != null || baseState.getSlot().equals(checkpoint.getEpochStartSlot())) {
```
I'm not sure why you would return a null pointer and then try to store that in checkpoint_states.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:423]: Actually I do see why you would check for null and return here, but in that scenario, we still have to deal with the null object in the caller method.

**Review by cemozerr — APPROVED** [review]
LGTM. Made one comment about returning a null pointer and placing that into a map.

File: storage/src/main/java/tech/pegasys/teku/storage/store/Store.java (hunk 1 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:423]: Good point, I've switched `regenerateCheckpointState` to return an `Optional` and flowed that around.

**Pull request merged**